File: leetcode/go/110.balanced-binary-tree.go
/*
 * @lc app=leetcode id=110 lang=golang
 *
 * [110] Balanced Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	left := 1 + maxDepth(root.Left)
	right := 1 + maxDepth(root.Right)
	if left > right {
		return left
	}
	return right
}
func isBalanced(root *TreeNode) bool {
	if root == nil {
		return true
	}

	if maxDepth(root.Left)-maxDepth(root.Right) >= -1 && maxDepth(root.Left)-maxDepth(root.Right) <= 1 {
		return isBalanced(root.Left) && isBalanced(root.Right)
	}
	return false
}

// @lc code=end

File: leetcode/cpp/110.balanced-binary-tree.cpp
/*
 * @lc app=leetcode id=110 lang=cpp
 *
 * [110] Balanced Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
 public:
  int getHeight(TreeNode* root) {
    if (root == nullptr) {
      return 0;
    }
    if (root->left == nullptr && root->right == nullptr) {
      return 1;
    }
    return 1 + max(getHeight(root->left), getHeight(root->right));
  }
  bool isBalanced(TreeNode* root) {
    if (root == nullptr) {
      return true;
    }
    int result = getHeight(root->left) - getHeight(root->right);
    if (result <= 1 && result >= -1) {
      return isBalanced(root->left) && isBalanced(root->right);
    }
    return false;
  }
};
// @lc code=end
